File: src/main/java/pt/isel/pc/sequences1/Sequence.java
// To access the *non-public* Continuation API
// ONLY for learning purposes
//@file:Suppress("JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")

package pt.isel.pc.sequences1;

import jdk.internal.vm.Continuation;
import jdk.internal.vm.ContinuationScope;

import java.util.Iterator;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.UnaryOperator;

public interface Sequence<T> extends Iterable<T> {
    
    static <T> Sequence<T> sequence(SequenceBlock<T> generator) {
        ContinuationScope scope = new ContinuationScope("SequenceScope");
        return () -> new SequenceIterator<>(generator, scope);
    }
    
    class SequenceIterator<T> implements  Iterator<T> {
        private T nextValue  = null;
        private Continuation cont;
        private ContinuationScope scope;
        
        public SequenceIterator(SequenceBlock<T> generator, ContinuationScope scope) {
            var t = this;
            cont = new Continuation(scope, () -> generator.exec(this));
            this.scope = scope;
        }
        
        @Override
        public boolean hasNext() {
            if (nextValue != null) return true;
            cont.run();
            return !cont.isDone() && nextValue != null;
        }
        
        @Override
        public T next() {
            if (!hasNext()) throw new IllegalStateException();
            var value = nextValue;
            nextValue = null;
            return value;
        }
        
        void yield(T value) {
            nextValue = value;
            Continuation.yield(scope);
        }
    }
   
    // generators
    
    static <T> Sequence<T> iterate(T initial, UnaryOperator<T> func) {
        return  sequence(s -> {
            T next = initial;
            while (true) {
                var n = next;
                next = func.apply(next);
                s.yield(n);
            }
        });
    }
    
    static Sequence<Integer>  range(int min, int max) {
        return sequence(seq -> {
            int curr = min;
            while (curr <= max) {
                seq.yield(curr);
                curr += 1;
            }
           
        });
    }
    
    // intermediate operations
    default Sequence<T> limit(int lim) {
        return  sequence(s -> {
            Iterator<T> srcIt = iterator();
            int curr = 0;
            
            while (curr++ < lim && srcIt.hasNext()) {
                s.yield(srcIt.next());
            }
          
        });
    }
    
    default <U,V>  Sequence<V> zip(Sequence<U> other , BiFunction<T,U,V> combiner )  {
        return  sequence(s -> {
            var itOther = other.iterator();
            var itSrc = iterator();
            while(itSrc.hasNext() && itOther.hasNext()) {
                s.yield(combiner.apply(itSrc.next(), itOther.next()));
            }
        });
    }
    
    default <U>  Sequence<U> flatMap(Function<T,Iterable<U>> mapper )  {
        return  sequence(s -> {
            for(var t : this) {
                for (var u : mapper.apply(t)) {
                    s.yield(u);
                }
            }
        });
    }
}

File: src/main/kotlin/pt/isel/pc/sequences/Sequence.kt
// To access the *non-public* Continuation API
// ONLY for learning purposes
@file:Suppress("JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")

package pt.isel.pc.sequences

import jdk.internal.vm.Continuation
import jdk.internal.vm.ContinuationScope

class Sequence<T>(val block : Sequence<T>.SequenceIterator.() -> Unit) : Iterable<T> {

    private val scope = ContinuationScope("SequenceScope")

    inner class SequenceIterator : Iterator<T> {
        private var nextValue : T? = null

        private val code : Runnable = Runnable {
            block.invoke(this)
        }

        private var cont = Continuation(scope, code)

        override fun hasNext(): Boolean {
            if (nextValue != null) return true
            cont.run()
            return !cont.isDone && nextValue != null
        }

        override fun next(): T {
            if (!hasNext()) throw IllegalStateException()
            val value = nextValue
            nextValue = null
            return value!!
        }

        fun yield(t: T) {
            nextValue = t
            Continuation.yield(scope)
        }
    }

    override fun iterator(): Iterator<T> {
        return SequenceIterator()
    }

    // generators

    companion object {
         fun <T> iterate(initial: T, func: ( (T) -> T))  =
             Sequence {
                 var next = initial
                 while(true) {
                     val n = next
                     next = func(next)
                     yield(n)
                 }
             }


        fun range(min: Int, max: Int): Sequence<Int> =
            Sequence {
                var curr = min
                while (curr <= max) {
                    yield(curr)
                    curr += 1
                }
            }
    }


    fun  limit(limit: Int) : Sequence<T> {
        val src = this
        return Sequence<T>()  {
            var curr = 0;
            val itSrc = src.iterator()
            while (curr++ < limit && itSrc.hasNext()) {
                yield(itSrc.next());
            }
        }
    }

    fun <U,V> zip(other : Sequence<U>, combiner: ((t: T, u: U) -> V) ): Sequence<V> {
        var src  = this
        return Sequence {
            val itOther = other.iterator()

            val itSrc = src.iterator()
            while(itSrc.hasNext() && itOther.hasNext()) {
                yield(combiner(itSrc.next(), itOther.next()));
            }
        }
    }
}